File: LeetCode Questions + Tutorials/EngineeringDigest/Leetcode1544.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Stack = /** @class */ (function () {
    function Stack() {
        this.items = [];
    }
    Stack.prototype.isEmpty = function () {
        return this.items.length === 0;
    };
    Stack.prototype.peek = function () {
        if (this.isEmpty()) {
            throw new Error("Stack is empty!");
        }
        else {
            return this.items[this.items.length - 1];
        }
    };
    Stack.prototype.add = function (elm) {
        this.items.push(elm);
    };
    Stack.prototype.pop = function () {
        if (this.isEmpty()) {
            throw new Error("Stack is empty!");
        }
        else {
            return this.items.pop();
        }
    };
    return Stack;
}());
function checkUpperCase(c) {
    return c === c.toUpperCase();
}
function checkLowerCase(c) {
    return c === c.toLowerCase();
}
function combineStackToString(stack) {
    var result = "";
    while (!stack.isEmpty()) {
        result = stack.pop() + result;
    }
    return result;
}
function makeGood(s) {
    var stack = new Stack();
    for (var _i = 0, s_1 = s; _i < s_1.length; _i++) {
        var c = s_1[_i];
        if (!stack.isEmpty()) {
            var topElm = stack.peek();
            if ((checkLowerCase(topElm) && checkUpperCase(c) && topElm.toLowerCase() === c.toLowerCase()) ||
                (checkUpperCase(topElm) && checkLowerCase(c) && topElm.toUpperCase() === c.toUpperCase())) {
                stack.pop();
                continue;
            }
        }
        stack.add(c);
    }
    return combineStackToString(stack);
}
;
var s = "leEeetcode";
console.log(makeGood(s));

File: LeetCode Questions + Tutorials/EngineeringDigest/Leetcode1544.java
import java.util.Scanner;
import java.util.Stack;

public class Leetcode1544 {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        System.out.print("Enter a string: ");
        String s = input.next();
        input.close();
        System.out.println(makeGood(s));
    }

    private static String makeGood(String s) {
        // ! Create a stack for string operations...
        Stack<Character> stack = new Stack<>();
        // * iterate every character */
        for (char c : s.toCharArray()) {
            // ! Check 1: Ensure the stack is not empty before proceeding.
            // ! Check 2: Evaluate the following conditions:
            //* a) The top value of the stack is lowercase, and the current character is
            //* uppercase,
            //* and both characters are the same when compared case-insensitively.
            //* OR
            //* b) The top value of the stack is uppercase, and the current character is
            //* lowercase,
            //* and both characters are the same when compared case-insensitively.

            if (!stack.isEmpty() &&
                    ((Character.isLowerCase(stack.peek()) && Character.isUpperCase(c)
                            && stack.peek() == Character.toLowerCase(c)) ||
                            (Character.isUpperCase(stack.peek()) && Character.isLowerCase(c)
                                    && stack.peek() == Character.toUpperCase(c)))) {
                stack.pop();
            } else {
                stack.add(c);
            }
        }
        StringBuilder result = new StringBuilder();
        for (char c : stack) {
            result.append(c);
        }
        return result.toString();
    }
}